juju_verify/verifiers/ceph.py

```diff
@@ -109,60 +137,83 @@ def __hash__(self) -> int:
-        parents_children_map = defaultdict(list)
+        if not all(self.get_node(name).type == "host" for name in names):
+            raise ValueError(
+                "Function can_remove_host_node is working only for node type host."
+            )
+
+        # Finds matching ancestors for host node.
+        ancestors_map = defaultdict(list)
         for name in names:
-            # NOTE (rgildein): `Self.get_node` could raise an error here, but the
+            # NOTE (rgildein): `self.get_node` could raise an error here, but the
             # check runner catches all exceptions.
-            children = self.get_nodes(name)
-            for child in children:
-                parent = self.find_parent(child.id)
-                if not parent:
-                    logger.warning(
-                        "A parent for the node %s could not be found.", child
-                    )
-                    return False
+            descendent = self.get_node(name)
+            ancestor = self.find_ancestor(descendent, required_ancestor_type)
+            if not ancestor:
+                raise ValueError(
+                    f"A ancestor for the host node {descendent} could not be found."
+                )
 
-                parents_children_map[parent].append(child)
+            ancestors_map[ancestor].append(descendent)
 
         # Check if all children could be removed from parent.
-        for parent, children in parents_children_map.items():
-            # NOTE (rgildein): This will check that the parent will have enough space
-            # even if the children are removed. An example with attempt to remove 2
-            # children:
+        for ancestor, descendents in ancestors_map.items():
+            # NOTE (rgildein): This will check that the ancestor will have enough space
+            # even if the descendent are removed. An example with attempt to remove 2
+            # descendents:
             #   parent with 5 children has 1 000 kB free space
             #   each child used the 400 kB space (2 000 kB total)
             #   each child has 200 kB of free space (1 000 kB total)
             #
             #   total available space after removing 2 units: 1 000kB - 2x200kB
             #   the total space that must moved to other units: 2x400kB
             #   check failed, due 600kB <= 800kB
-            total_children_kb_used = sum(child.kb_used for child in children)
-            total_children_kb_avail = sum(child.kb_avail for child in children)
-            if (parent.kb_avail - total_children_kb_avail) <= total_children_kb_used:
+            total_descendent_kb_used = sum(
+                descendent.kb_used for descendent in descendents
+            )
+            total_descendent_kb_avail = sum(
+                descendent.kb_avail for descendent in descendents
+            )
+            if (
+                ancestor.kb_avail - total_descendent_kb_avail
+            ) <= total_descendent_kb_used:
```

Comment:
This looks good. We look for the total amount of available space removing the info from the unit(s) we want to remove. With this, we get the real available space and we compare it with the amount of space used by the unit(s) that will be rebooted :+1: